iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -625,27 +741,17 @@ static LogicalResult setRootConfig(
       iterationDomain, partitionableLoopsInterfaceOp, minTileSizes,
       maxTileSizes);
 
-  // Set the Next level tile sizes.
-  SmallVector<int64_t> parallelTileSizes(numLoops, 0);
-  Optional<SmallVector<int64_t, 4>> staticLoopRanges =
-      linalgOp.getStaticLoopRanges();
-  for (auto loopNum : llvm::seq<unsigned>(0, numLoops)) {
-    if (flowTileSizes[loopNum]) {
-      parallelTileSizes[loopNum] =
-          getMaxTileSize(0, flowTileSizes[loopNum], minTileSizes[loopNum],
-                         minTileSizes[loopNum]);
-    } else {
-      // If the flow level tile size is zero, and static loop range is 0 as
-      // well, set the tile sizes here to zero as well.
-      parallelTileSizes[loopNum] =
-          (staticLoopRanges && staticLoopRanges.getValue()[loopNum] == 1)
-              ? 0
-              : minTileSizes[loopNum];
-    }
-  }
+  // Set the next level tile sizes.
+  SmallVector<int64_t> parallelTileSizes;
   SmallVector<int64_t> reductionTileSizes;
-  splitParallelAndReductionTiles(linalgOp, parallelTileSizes,
-                                 reductionTileSizes);
+  if (failed(setX86TransposeParallelReductionTileSizes(
+          genericOp, numLoops, flowTileSizes, minTileSizes, maxTileSizes,
+          parallelTileSizes)))
+    return failure();
+  if (failed(setX86DefaultParallelReductionTileSizes(
+          genericOp, numLoops, flowTileSizes, minTileSizes, maxTileSizes,
+          parallelTileSizes, reductionTileSizes)))
+    return failure();
```

Comment:
Sorry, let me give one step back here. I think I broke the functionality while addressing the feedback! Currently both `setX86...` utilities above are executing and modifying the `parallelTileSizes` for the transpose cases, which is incorrect. `setX86DefaultParallelReductionTileSizes` is checking `if (getLoweringConfig(genericOp))` to proceed or not, and that check is always false because `setX86TransposeParallelReductionTileSizes` is actually not setting the configuration (that is happening a few lines below (766)). 

The fact that both utilities are always returning `success` is not making too much sense either. If we don't want to return `failure` because that is reserved for other situations, I don't think we can avoid returning a bool and an if-else statement here (that could eventually turn into a switch, if we had more op specific cases). If you feel strong against this, please let me know the alternative.

Sorry, my fault. I think I've been following the feedback without thinking too much about it.

Reply:
Sorry, I didnt read through the logic fully. At this level maybe making a `bool` and `if-else` is fine. If I were to write this I would structure this differently. I would basically treat this as a "new root op configuration setting method". So use the `setRootConfig` as just a point to trampoline into two separate methods that set the configuration on success if that was the intent. That is much cleaner for me to read. But in this case `bool` + `if-else` is also fine..

Reply:
Ok, looking through with fresher eyes, here is how I would do this. I would rewrite this method to 
```
LogicalResult setRootConfig(func::FuncOp .., linalg::GenericOp genericOp ...) {
  if (failed(setTransposeLikeOpRootConfig(..., genericOp, ...) ||
      failed(setDefaultGenericOpRootConfig(...., genericOp,....)) {
      return failure();
   }
   return success();
}
```

and then each of those methods

```
LogicalResult setTransposeLikeOpRootConfig(...., linalg::GenericOp genericOp ) {
   if (getLoweringConfig(genericOp)) return succes();
   ....
}

LogicalResult setDefaultGenericOpRootConfig(..., linalg::GenericOp genericOp) {
   if (getLoweringConfig(genericOp)) return succes();
   ....

}
```

This is how I set this configuration setting up. THe logic here can get very complicated with `if-else-if-else` and branching. So this approach decouples all the `set*RootConfig` methods. I'll let Hanhan and you decide if you want to keep this flow. The worst outcome would be to mix-and-match. If you guys decide to change the flow in this file, just make the whole thing consistent.

Reply:
This sounds interesting! Thanks! Looking at the other `setRootConfig` methods I see that many are duplicating similar code so I guess that's fine for now. It's not clear, though, how this is going to compose because a transpose op it's just a generic op that may have other ops inside that may change the configuration in the future. I guess we can think about that later.

Let me give this a try!

Reply:
Well, I am not super tied to this approach. I have had the misfortune of having to go deep into this file a couple of times, and the only way I could make it managable for me is to essentially have all `setRoot*Config` layered this way. I dont anticipate touching these files as much (at least not as much as you and Hanhan). So I am completely open to reworking the way this is setup as long as the approach is made consistent. A mix and match of things is where things get really confusing.

In this particular case, I agree that the transpose op being treated specially from other generic ops is kind of an issue. I'd say this is a temporary state. It is very rare you will see generic ops with just one input and one output at this stage. So this is targeting a very specific pattern. Ideally in the end state the separate logic for transpose lowering will just get included into the default generic op configuration, but I realize there is some work to get there.

Reply:
Thanks for the feedback! Now that I have a more global understanding things make more sense to me. I'm aligned with your thinking above. Regarding how the generic op logic will evolve, I have the impression that we will need a more sophisticated "matching" phase to get more meaningful information out of it and then dispatch to a specific configuration. However, I think we need to better understand the use cases before making any significant design change.

Hopefully, the last version makes sense for now! Ptal.

Reply:
This makes sense to me. I see my stamp went away :) . Sorry didnt realize. Stamping again. Thanks for working with me on this. Just restating to emphasize, I am not tied to this flow, more what made sense to me when I did deep dive into it. If there is a better way I am all for it!